ntoskrnl/fsrtl/fastio.c

```diff
@@ -1552,46 +1552,63 @@ FsRtlPrepareMdlWriteDev(IN PFILE_OBJECT FileObject,
-    FilterCallbacks = DriverExtension->FsFilterCallbacks;
+    switch (Operation)
+    {
+        case FS_FILTER_ACQUIRE_FOR_SECTION_SYNCHRONIZATION:
+            return RTL_SIZEOF_THROUGH_FIELD(FS_FILTER_CALLBACKS, PostAcquireForSectionSynchronization);
+        case FS_FILTER_RELEASE_FOR_SECTION_SYNCHRONIZATION:
+            return RTL_SIZEOF_THROUGH_FIELD(FS_FILTER_CALLBACKS, PostReleaseForSectionSynchronization);
+        case FS_FILTER_ACQUIRE_FOR_CC_FLUSH:
+            return RTL_SIZEOF_THROUGH_FIELD(FS_FILTER_CALLBACKS, PostAcquireForCcFlush);
+        case FS_FILTER_RELEASE_FOR_CC_FLUSH:
+            return RTL_SIZEOF_THROUGH_FIELD(FS_FILTER_CALLBACKS, PostReleaseForCcFlush);
+        case FS_FILTER_ACQUIRE_FOR_MOD_WRITE:
+            return RTL_SIZEOF_THROUGH_FIELD(FS_FILTER_CALLBACKS, PostAcquireForModifiedPageWriter);
+        case FS_FILTER_RELEASE_FOR_MOD_WRITE:
+            return RTL_SIZEOF_THROUGH_FIELD(FS_FILTER_CALLBACKS, PostReleaseForModifiedPageWriter);
+        default:
+            return 0;
+    }
+}
 
-    /* Check if Filter Cllbacks are supported */
-    if (FilterCallbacks && FilterCallbacks->PreAcquireForSectionSynchronization)
+static
+NTSTATUS
+FsRtlpCallPrePostCallbacks(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PFILE_OBJECT FileObject,
+    _In_ UCHAR Operation,
+    _In_ PFS_FILTER_CALLBACKS FilterCallbacks,
+    _In_opt_ PFS_FILTER_PARAMETERS FilterParameters,
+    _In_ PFS_FILTER_CALLBACK PreCallback,
+    _In_ PFS_FILTER_COMPLETION_CALLBACK PostCallback)
+{
+    /* Check if Filter Callbacks are supported */
+    if (FilterCallbacks &&
+        FilterCallbacks->SizeOfFsFilterCallbacks >= FsRtlpGetPostCallbackSize(Operation) &&
+        DeviceObject != IoGetBaseFileSystemDeviceObject(FileObject) &&
+        PreCallback)
     {
         NTSTATUS Status;
         PVOID CompletionContext;
 
-        FS_FILTER_CALLBACK_DATA CbData;
-
-        RtlZeroMemory(&CbData, sizeof(CbData));
-
-        CbData.SizeOfFsFilterCallbackData = sizeof(CbData);
-        CbData.Operation = FS_FILTER_ACQUIRE_FOR_SECTION_SYNCHRONIZATION;
+        FS_FILTER_CALLBACK_DATA CbData = { sizeof(CbData) };
+        CbData.Operation = Operation;
         CbData.DeviceObject = DeviceObject;
         CbData.FileObject = FileObject;
-        CbData.Parameters.AcquireForSectionSynchronization.PageProtection = Reserved;
-        CbData.Parameters.AcquireForSectionSynchronization.SyncType = SyncType;
 
-        Status = FilterCallbacks->PreAcquireForSectionSynchronization(&CbData, &CompletionContext);
+        if (Operation == FS_FILTER_ACQUIRE_FOR_SECTION_SYNCHRONIZATION ||
+            Operation == FS_FILTER_ACQUIRE_FOR_MOD_WRITE ||
+            Operation == FS_FILTER_RELEASE_FOR_MOD_WRITE)
+        {
+            CbData.Parameters = *FilterParameters;
+        }
+
+        Status = PreCallback(&CbData, &CompletionContext);
         if (!NT_SUCCESS(Status))
         {
             FsRtlExitFileSystem();
```

Comment:
Why is this calling `FsRtlExitFileSystem`?

Reply:
Because when entering any of the acquire/release FsRtl routines, we call `FsRtlEnterFileSystem()` at the start, before doing anything else, and hence according to the correct logic, we need to call an appropriate routine which reverts the thing in a failure case. And Windows seems to do this too in a failure case as well. But if you have any arguments against and/or this is not actually required, I can remove it if needed. Also, it probably shouldn't be executed when is called from `FsAcquire`/`ReleaseFileForModWrite`(`Ex`), since these two routines don't call `FsRtlEnterFileSystem()` unlike others. So it can be limited to other 4 routines only.